openedx/core/djangoapps/course_groups/models.py

```diff
@@ -58,6 +61,112 @@ def create(cls, name, course_id, group_type=COHORT):
+    #for django 1.8 upgrade, just remove this method and allow the base method to be called instead.
+    #reference: https://docs.djangoproject.com/en/1.6/ref/models/instances/, under "Validating Objects"
+    def full_clean(self, **kwargs):
+        self.clean_fields()
+        self.clean()
+        if 'validate_unique' not in kwargs or kwargs['validate_unique'] is True:
+            self.validate_unique()
+
+    def clean_fields(self, *args, **kwargs):
+        if self.course_id is None:
+            self.course_id = self.course_user_group.course_id
+        super(CohortMembership, self).clean_fields(*args, **kwargs)
+
+    def clean(self):
+        if self.course_user_group.group_type != CourseUserGroup.COHORT:  # pylint: disable=E1101
+            raise ValidationError("CohortMembership cannot be used with CourseGroup types other than COHORT")
+        if self.course_user_group.course_id != self.course_id:
+            raise ValidationError("Non-matching course_ids provided")
+
+    def save(self, *args, **kwargs):
+        self.full_clean(validate_unique=False)
+
+        #print "Trying to add user to {}".format(self.course_user_group.name)
+
+        trying_to_save = True
+        while trying_to_save:
+            """print "entering loop, trying to get membership for user {user}, course {course}".format(
+                user=self.user.id,  # pylint: disable=E1101
+                course=self.course_id
+            )"""  # pylint: disable=W0105
+
+            saved_membership = None
+            try:
+                #the following 2 "transaction" lines force a fresh read, they can be removed once we're on django 1.8
+                #http://stackoverflow.com/questions/3346124/how-do-i-force-django-to-ignore-any-caches-and-reload-data
+                with transaction.commit_manually():
+                    transaction.commit()
+
+                saved_membership = CohortMembership.objects.get(
+                    user__id=self.user.id,  # pylint: disable=E1101
+                    course_id=self.course_id
+                )
+            except CohortMembership.DoesNotExist:
+                #print "No saved_membership!"
+                pass
+
+            if saved_membership is not None:
+                if saved_membership.course_user_group == self.course_user_group:
+                    raise ValueError("User {user_name} already present in cohort {cohort_name}".format(
+                        user_name=self.user.username,  # pylint: disable=E1101
+                        cohort_name=self.course_user_group.name
+                    ))
+                self.previous_cohort = saved_membership.course_user_group
+                self.previous_cohort_name = saved_membership.course_user_group.name
+                self.previous_cohort_id = saved_membership.course_user_group.id
+
+            time.sleep(10)
+
+            try:
+                with transaction.commit_on_success():
```

Comment:
There are 2 other points in this PR that explicitly need changes to work on django 1.8.

https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/models.py#L70
and
https://github.com/edx/edx-platform/blob/efischer/cohort_membership_model/openedx/core/djangoapps/course_groups/models.py#L104

They're both fairly simple changes that I tried to outline with inline comments. I'd be glad to help get this code up-to-date when the time comes.